mne/tests/test_morph.py

```diff
@@ -0,0 +1,438 @@
+
+# Setup paths
+tempdir = _TempDir()
+
+data_path = testing.data_path()
+
+sample_dir = op.join(data_path, 'MEG', 'sample')
+
+subjects_dir = op.join(data_path, 'subjects')
+
+fname_evoked = op.join(sample_dir, 'sample_audvis-ave.fif')
+
+fname_fwd_vol = op.join(sample_dir,
+                        'sample_audvis_trunc-meg-vol-7-fwd.fif')
+fname_inv_vol = op.join(sample_dir,
+                        'sample_audvis_trunc-meg-vol-7-meg-inv.fif')
+fname_vol = op.join(sample_dir,
+                    'sample_audvis_trunc-grad-vol-7-fwd-sensmap-vol.w')
+fname_inv_surf = op.join(sample_dir,
+                         'sample_audvis_trunc-meg-eeg-oct-6-meg-inv.fif')
+
+fname_smorph = op.join(sample_dir, 'sample_audvis_trunc-meg')
+fname_fmorph = op.join(sample_dir, 'fsaverage_audvis_trunc-meg')
+
+fname_t1 = op.join(subjects_dir, 'sample', 'mri', 'T1.mgz')
+
+
+def _real_vec_stc():
+    inv = read_inverse_operator(fname_inv_surf)
+    evoked = read_evokeds(fname_evoked, baseline=(None, 0))[0].crop(0, 0.01)
+    return apply_inverse(evoked, inv, pick_ori='vector')
+
+
+@requires_nibabel()
+@testing.requires_testing_data
+def test_save_vol_stc_as_nifti():
+    """Save the stc as a nifti file and export."""
+    import nibabel as nib
+    tempdir = _TempDir()
+    with warnings.catch_warnings(record=True):
+        warnings.simplefilter('always')
+        src = read_source_spaces(fname_fwd_vol)
+    vol_fname = op.join(tempdir, 'stc.nii.gz')
+
+    # now let's actually read a MNE-C processed file
+    stc = read_source_estimate(fname_vol, 'sample')
+    assert (isinstance(stc, VolSourceEstimate))
+
+    stc.save_as_volume(vol_fname, src,
+                       dest='surf', mri_resolution=False)
+    with warnings.catch_warnings(record=True):  # nib<->numpy
+        img = nib.load(vol_fname)
+    assert (img.shape == src[0]['shape'] + (len(stc.times),))
+
+    with warnings.catch_warnings(record=True):  # nib<->numpy
+        t1_img = nib.load(fname_t1)
+    stc.save_as_volume(op.join(tempdir, 'stc.nii.gz'), src,
+                       dest='mri', mri_resolution=True)
+    with warnings.catch_warnings(record=True):  # nib<->numpy
+        img = nib.load(vol_fname)
+    assert (img.shape == t1_img.shape + (len(stc.times),))
+    assert_allclose(img.affine, t1_img.affine, atol=1e-5)
+
+    # export without saving
+    img = stc.as_volume(src, dest='mri', mri_resolution=True)
+    assert (img.shape == t1_img.shape + (len(stc.times),))
+    assert_allclose(img.affine, t1_img.affine, atol=1e-5)
+
+    src = SourceSpaces([src[0], src[0]])
+    stc = VolSourceEstimate(np.r_[stc.data, stc.data],
+                            [stc.vertices, stc.vertices],
+                            tmin=stc.tmin, tstep=stc.tstep, subject='sample')
+    img = stc.as_volume(src, dest='mri', mri_resolution=False)
+    assert (img.shape == src[0]['shape'] + (len(stc.times),))
+
+    assert_raises(ValueError, stc.as_volume, src, mri_resolution=(4., 4., 4.))
+
+
+@pytest.mark.slowtest
```

Comment:
how long do the tests take?

Reply:
my bench says:
29.64s call     mne/tests/test_morph.py::test_volume_source_morph
13.96s call     mne/tests/test_morph.py::test_morph_data
2.70s call     mne/tests/test_morph.py::test_surface_vector_source_morph
0.44s call     mne/tests/test_morph.py::test_save_vol_stc_as_nifti
0.39s call     mne/tests/test_morph.py::test_stc_as_volume

Reply:
okay you might want to bring it down to 10 seconds. 30 seconds is huge !

Reply:
hm... so splitting it into sub-tests?

Reply:
haha no. I mean actually making it faster. Maybe try profiling your code and seeing what takes time. Try to use fewer channels / source points. Decimated grid, less iterations. There are different tricks you can use to make things faster

Reply:
maybe see what you can precompute and load the precomputed file.

Reply:
the long lasting stuff is the computation of the non-linear morph... I see how much I can strip it - otherwise I already load everything

Reply:
guess that's all I can offer:
13.82s call     mne/tests/test_morph.py::test_morph_data
10.08s call     mne/tests/test_morph.py::test_volume_source_morph
2.62s call     mne/tests/test_morph.py::test_surface_vector_source_morph
0.47s call     mne/tests/test_morph.py::test_save_vol_stc_as_nifti
0.38s call     mne/tests/test_morph.py::test_stc_as_volume